### ISSUE TITLE
Interoperable Private Attribution (IPA)

### ISSUE THREAD

**Issue by eriktaubeneck** [patcg/proposals#2]
@benjaminsavage, @martinthomson, and I have been working on a proposal, "[Interoperable Private Attribution (IPA)](https://docs.google.com/document/d/1KpdSKD8-Rn0bWPTu4UtK54ks0yv2j22pA5SrAD9av4s/edit#)" that addresses the aggregate attribution measurement use case, similar to those listed in #1. 

We'd love to have this considered and discussed at the January PATCG meeting, for consideration in maturing it further through collaboration among this community group.

**Comment by tgeoghegan** [comment]
[Section 4.1](https://docs.google.com/document/d/1KpdSKD8-Rn0bWPTu4UtK54ks0yv2j22pA5SrAD9av4s/edit?pli=1#heading=h.s1l843iiwg07) states that the "aim is for IPA to be compatible with the Privacy Preserving Measurement (PPM) specification". Does that mean you intend to express IPA as a [VDAF](https://cjpatton.github.io/vdaf/draft-patton-cfrg-vdaf.html)?

**Comment by eriktaubeneck** [comment]
> Does that mean you intend to express IPA as a VDAF?
Yes, our intention is to work towards that. There are a few major components, some which likely require expression as a new VDAF, but we hope to leverage the existing work with prio3 and/or poplar1 where possible.

**Comment by alextcone** [comment]
Consider me a strong +1 in support of this getting on the agenda.

**Comment by ekr** [comment]
In case it helps, I spent a bunch of time working through the math of IPA in some detail at: https://educatedguesswork.org/posts/vaccine-tracking/. I was interested in another application, but if you found the ElGamal blinding and shuffling a bit had, this might help.

**Comment by ansuz** [comment]
It seems like permissions are required to view the draft on google docs. Is a publicly accessible version of the draft available anywhere else?

**Comment by gsnedders** [comment]
While the proposal mentions:

> We would also like to call out the work happening in the WICG on the [Attribution Reporting API](https://github.com/WICG/conversion-measurement-api) and [Privacy Preserving Ads](https://github.com/WICG/privacy-preserving-ads), which was highly influential to this work.

It would be nice for the proposal to directly compare itself with such prior work (and also the Privacy CG's [Private Click Measurement](https://github.com/privacycg/private-click-measurement)). As it is, it is not immediately clear as to what the motivations are behind this proposal rather than furthering work on developing the other proposals.

**Comment by ShivanKaul** [comment]
Is there a reason all the questions on the doc were removed + ability to comment revoked?

**Comment by Lexicality** [comment]
Presumably because the internet is currently extremely angry about this?

**Comment by benjaminsavage** [comment]
> Is there a reason all the questions on the doc were removed + ability to comment revoked?

The document was completely defaced, fully deleted with "suggestions" and replaced with vulgarities. As such, the document is now "read-only" access.

**Comment by ShivanKaul** [comment]
Ah, I see, sorry to hear that. Is the plan to move it to GitHub? Last time I read it there were a few undefined terms and the flow was not entirely clear to me, would be good to get clarifying answers.

**Comment by bmayd** [comment]
> The document was completely defaced, fully deleted with "suggestions" and replaced with vulgarities. As such, the document is now "read-only" access.

@benjaminsavage Does this lead you to consider docs unsuitable as a collaboration tool for this work or do you think it can be avoided going forward? I don't want to continue advocating for their use if the latter is not the case.

**Comment by medicinalcocaine3434** [comment]
> Does this lead you to consider docs unsuitable as a collaboration tool for this work or do you think it can be avoided going forward? I don't want to continue advocating for their use if the latter is not the case.
[this might be a solution](https://support.google.com/docs/answer/6033474)

**Comment by santirely** [comment]
Having looked at only the non-technical presentation, I have a couple of comments / questions.

1. Where would the match keys be stored in the device / browser / OS?
2. Could we adapt this solution so that it isn't so dependent on companies with a "large footprint"? Although as proposed this does appear to be the most logical solution, it also poses a huge limitation in my view. What happens if suddenly Facebook or Google decide they'll only share their match keys with companies that play well with them?

**Comment by eriktaubeneck** [comment]
@medicinalcocaine3434 the document was using suggested edits and comments, it was with suggested edits that the document was defaced.

@santirely if you take a look at the technical proposal, you'll find details on your question. Briefly:
> Where would the match keys be stored in the device / browser / OS?

We are proposing a new _read-only_ API, which the browser/OS would expose.

> Could we adapt this solution so that it isn't so dependent on companies with a "large footprint"?

Any website/app is able to write a match key, so it's not dependent on any set of companies. However, the more cross-device coverage a given companies match key has, the more accurate attribution that uses that match key will be.

> What happens if suddenly Facebook or Google decide they'll only share their match keys with companies that play well with them?

We are proposing that any site/app be able to reference any match key. Match keys are not shared, and the ability to reference them is not controlled by the companies that set them.

**Comment by santirely** [comment]
Ok, that's interesting. A couple follow ups:

> Any website/app is able to write a match key, so it's not dependent on any set of companies. However, the more cross-device coverage a given companies match key has, the more accurate attribution that uses that match key will be.

This is true but a significant portion of the value created by the proposal is cross-device tracking, and these companies adopting the solution would be important for that to actually work. I was aiming at something like: What if other smaller apps / publishers could pool their match keys in a way that benefits everyone? For example, a gaming studio like say Epic will have tons of mobile and CTV match keys but almost no web based ones. Opposite is true for someone like The New York Times. Could they create a sort of coop there?

> What happens if suddenly Facebook or Google decide they'll only share their match keys with companies that play well with them?

This seems ideal, but isn't that a potential drawback for someone that can provide cross-device atribution by itself? What's Meta's or Google's incentive to be the world's match key providers there?

**Comment by Lexicality** [comment]
The proposal also seems to assume that the user is logged in to at least one match key provider. What happens if the user is not? Does the browser make up a device identifier? Does it encrypt null? (Presumably collisions could abound there) or does it return an error to the calling script?
If a user doesn't want to sign in to Facebook (or uses [Facebook Container](https://addons.mozilla.org/en-GB/firefox/addon/facebook-container/)) does that mean they will never be able to be attributed to a conversion?

**Comment by eriktaubeneck** [comment]
> The proposal also seems to assume that the user is logged in to at least one match key provider. What happens if the user is not? Does the browser make up a device identifier?

This is still yet to be determined. One idea is that the device could generate a random match key, which would at least default to "same device attribution".

> If a user doesn't want to sign in to Facebook (or uses [Facebook Container](https://addons.mozilla.org/en-GB/firefox/addon/facebook-container/)) does that mean they will never be able to be attributed to a conversion?

This would entirely depend which match key providers the source sites and trigger sites are using. If those sites are using match key providers that the user is not logged into, then they attribution would likely be missed. 100% coverage have never been possible, but this API is designed to create as much coverage as possible, without enabling user level tracking.

**Comment by Lexicality** [comment]
If Twitter wants to use Facebook's match keys, that means all the users they show adverts to also need to be logged in to Facebook. This means Twitter needs to incentivise its users to log in to Facebook which directly benefits Facebook.

That's the entire reason Facebook has come up with this proposal - in order for it to work the vast majority of the internet needs to have been identified in some fashion by Facebook, so everyone that uses it will pass their users via Facebook and let them slurp up their data.

You can say "well it works with any provider" but to work effectively it needs a major provider and Google are off doing their own thing so ...

**Comment by bmilekic** [comment]
> If Twitter wants to use Facebook's match keys, that means all the users they show adverts to also need to be logged in to Facebook. This means Twitter needs to incentivise its users to log in to Facebook which directly benefits Facebook.

I don't see how that's true. If Twitter is the publisher, then it can ask their advertisers to register _trigger_ events referencing only twitter's match keys. There is no need for the facebook match keys in that scenario, especially since the ads are running on twitter and so the lack of twitter user ID implies no possible match with advertiser _target_ events.

For non-FB smaller publishers, the proposal provides an important theoretical benefit, in that a publisher can choose to register _source_ events leveraging facebook, twitter, and other third-party keys. In an IPA implementation supporting multiple match keys, this ultimately benefits the publisher as it increases potential match rates with advertiser data.

It remains to be seen what the motivation could be for a match key provider to act as such, but presumably "having an advertising business" would be one motivating reason. I believe that making the match keys usable by other parties in that context is more fair than doing the opposite.

**Comment by Lexicality** [comment]
> I don't see how that's true. If Twitter is the publisher, then it can ask their advertisers to register trigger events referencing only twitter's match keys. There is no need for the facebook match keys in that scenario, especially since the ads are running on twitter and so the lack of twitter user ID implies no possible match with advertiser target events.

I agree it doesn't make much sense, but in the hypothetical that they did want to only use someone else's match key (for whatever reason) then I think my point still stands

> It remains to be seen what the motivation could be for a match key provider to act as such, but presumably "having an advertising business" would be one motivating reason. I believe that making the match keys usable by other parties in that context is more fair than doing the opposite.

I don't think fairness comes in to many business decisions. It costs Facebook nothing to allow its competitors to use its match keys, and if everyone relies on them they gain a position of power over the discourse, even if it's just an implicit one.

To be clear, just because I feel like this proposal further entrenches the big players in the ad business by relying on centralised identity services doesn't mean I think it's a _bad_ proposal. As long as the cryptographic stuff works and the ad networks are somehow coerced into dropping their other tracking methods, this is a big step up. But on the other hand if the crypto stuff has a hidden weakness in it and Facebook run one of the "trusted" servers, this is a terrible idea.

**Comment by benjaminsavage** [comment]
A few comments in response the the thread so far:

1. On the topic of "what if no match keys are set"

As @eriktaubeneck said - I like the idea of the device just generating a random matchkey. That way the API seamlessly defaults to "same-device-only" attribution, which is at least at par with other proposals.

2. On the topic of "Who can use a match key once it is set?"

The reason we proposed allowing any company to benefit from match keys set by any other participant, was specifically to try to avoid any kind of system which could be abused by large established players. As @[santirely](https://github.com/santirely) mentions, this would give them a lot of leverage to, as he says, choose to only share access with businesses who "play well with them". We opted for an "open reference" proposal specifically to avoid this type of risk.

3. On the topic of: "What incentive would a company with a large footprint have for setting an open-reference match key?"

As @eriktaubeneck points out, browsers and mobile operating systems are rapidly clamping down on "tracking". Various regulations are doing the same. This means that all businesses (even those with a large footprint) as steadily losing the ability to accurately count the number of conversions attributable to advertising. In a theoretical future world where cookies and device identifiers are all gone, and fingerprinting is impossible, having a "large footprint" will be _useless_ from the perspective of counting conversions which occur off-network on other apps and websites. In such a world, if the only option available for counting conversions is a highly private one, like IPA, then I believe businesses who sell ads will use it (they won't have a choice). In that world, they'll have two options:
  (i) Do not set a match key. Use a match key set by some other entity
  (ii) Set a match key - accepting that anyone else who wants to can also use it.
  
Each entity will have to weigh these alternatives. For a business with a "large footprint" of users who sign in across multiple devices, here is how I think these choices will look:
  (i) **Do not set a match key:** If other match-keys are from businesses with a smaller network of users logged-in across devices, taking this approach will have the un-desireable side-effect of *undercounting* the true number of conversions their ads actually drive. In summary: Less accurate measurement.
  (ii) **Set a match key:** This will result in more accurate ads measurement - with higher counts of attributed conversions, which more accurately measures the number of conversions their ads drive. As a side-effect however, all competitors will **also** benefit from more accurate measurement of their ads. In summary: More accurate, but more accurate for everyone.
  
I posit that there exist businesses for whom the calculus is in favor of option (ii), more accurate measurement being more beneficial than everyone having less accurate measurement.

4. On the topic of "does this require users to be logged into Facebook?". In the proposal, we talk about the prospect of supporting **multiple match keys**. We think we can support this without needing to give up any privacy benefits. If that is true, then it would seem optimal for any consumer of this API to select a basket of match-keys which collectively provide good coverage. This has the additional benefit of minimizing the reliance on a single point of failure. I can envision a future where it is common to specify a handful of "large footprint" match key providers to get a good baseline, a few region specific ones to cover parts of the globe which would otherwise be poorly covered, potentially one's own match-key, and finally falling back on the random, per-device specified match key which essentially just provides "same-device only" attribution. 

I think all parties (including "large footprint" entities) would all have similar incentives to push them in this fashion.

We've also put a lot of time and thought into trying to ensure there isn't coupling between entities. We think we can design the system in such a way that we do not require collaboration. That is, we want a system where any advertiser who runs ads across N platforms can **independently** specify which match-keys they want to use, without needing those platforms to all agree with them, or all need agree on something.

5.
> As long as the cryptographic stuff works and the ad networks are somehow coerced into dropping their other tracking methods, this is a big step up. But on the other hand if the crypto stuff has a hidden weakness in it and Facebook run one of the "trusted" servers, this is a terrible idea

First of all, I assume that Facebook / Google / any ad-tech company will never be trusted to operate a helper server =). This will be enforced by browsers. They'll have to decide which public keys they are willing to use to encrypt reports. I cannot imagine a world in which Firefox would trust Facebook enough to encrypt these events using Facebook's public key =). I'm assuming we will see non-profits with strong privacy reputations operating the servers, or possibly the types of organizations which operate Apple's "Private Relay" service.

Secondly: Yes, exactly. This proposed system would be a **big** step up for privacy compared to the status quo mechanisms used to count conversions. I have no expectation that browsers and mobile operating systems will stop trying to clamp down on fingerprinting. Actually, if anything I expect them to accelerate those efforts. I also expect to see more and more regulation along these lines. 

That the math works out, and we have a strong privacy guarantee is the key. This is why we are trying to work out in the open - we think that's the best way to find all the problems / issues, and to get help finding solutions to them. We've already benefitted tremendously from outside input. @betuldurak found a really clever attack that a malicious helper node could do. I'm really grateful to her for telling us about it! We're working on finding a solution as we speak. 

I think the path towards standardization looks like a bunch of iterations out in the open, publishing papers, getting feedback, addressing problems, repeat. I hope that we can eventually converge on a design that is super solid. I wouldn't expect browser vendors to feel comfortable shipping an API like this unless a bunch of independent academics were all convinced that it met our design goals.

**Comment by martinthomson** [comment]
https://educatedguesswork.org/posts/ipa-overview/#appendix%3A-linear-relation-attacks perhaps.

We've initiated a few discussions with cryptographers; nothing public as yet.

**Comment by sthaase** [comment]
What role do regulatory requirements such as GDPR / ePrivacy in Europe play in the solution discovery & design from your perspective? That is one aspect I rarely read about in these proposals, yet I believe that this should be an integral part of the problem definition and solution design. 

Looking at IPA specifically for example, I believe that data protection authorities might categorize the match key as personal data (https://gdpr.eu/eu-gdpr-personal-data/) and storing it on the users device would therefore require user consent. Would you just accept that as a given, or could solutions be more tailored towards regulatory requirements (in a sense: try to discover solutions do not require user consent to not end up modeling 30% of conversions that are lost due to tracking opt-outs).

**Comment by benjaminsavage** [comment]
> What role do regulatory requirements such as GDPR / ePrivacy in Europe play in the solution discovery & design from your perspective? That is one aspect I rarely read about in these proposals, yet I believe that this should be an integral part of the problem definition and solution design.

@alextcone and @darobin shared some of their thoughts about this topic in another thread:

@alextcone's comment: https://github.com/patcg/proposals/issues/5#issuecomment-1034987044

@darobin's comment: https://github.com/patcg/proposals/issues/5#issuecomment-1035273304

**Comment by AramZS** [comment]
Noting that the group has chosen to pick up private measurement and the associated proposals as part of its initial focus. This proposal will be considered as part of that process. Accordingly, I am moving this issue to the stand alone repo created to manage that conversation.
